// create-app/lib/share/createMatcher.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var path_to_regexp_1 = __importDefault(require("path-to-regexp"));
function createMatcher(routes) {
    var finalRoutes = routes.map(createRoute);
    var routeLength = finalRoutes.length;
    var matcher = function (pathname) {
        var finalPathname = cleanPath(pathname);
        for (var i = 0; i < routeLength; i++) {
            var route = finalRoutes[i];
            var strMatches = route.regexp.exec(finalPathname);
            if (!strMatches) {
                continue;
            }
            var params = getParams(strMatches, route.keys);
            var controller = route.controller;
            return {
                path: route.path,
                params: params,
                controller: controller
            };
        }
        return null;
    };
    return matcher;
}
exports.default = createMatcher;
function createRoute(route) {
    var finalRoute = Object.assign({}, route);
    finalRoute.keys = [];
    var keys = finalRoute.keys;
    var regexp = path_to_regexp_1.default(finalRoute.path, keys);
    var intactRoute = Object.assign({ keys: keys, regexp: regexp }, finalRoute);
    return intactRoute;
}
function getParams(matches, keys) {
    var params = {};
    for (var i = 1, len = matches.length; i < len; i++) {
        var key = keys[i - 1];
        if (key && matches[i]) {
            if (typeof matches[i] === 'string') {
                params[key.name] = decodeURIComponent(matches[i]);
            }
            else {
                params[key.name] = matches[i];
            }
        }
    }
    return params;
}
function cleanPath(path) {
    return path.replace(/\/\//g, '/');
}
