// create-history/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
}
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMemoryHistory = exports.createHashHistory = exports.createHistory = void 0;
///////////////////////////////////////////////////////////////////////////////
// FUNCTIONS
///////////////////////////////////////////////////////////////////////////////
var createBrowserHistory_1 = __importDefault(require("./createBrowserHistory"));
var createHashHistory_1 = __importDefault(require("./createHashHistory"));
var createMemoryHistory_1 = __importDefault(require("./createMemoryHistory"));
exports.createHistory = createBrowserHistory_1.default;
exports.createHashHistory = createHashHistory_1.default;
exports.createMemoryHistory = createMemoryHistory_1.default;
exports.default = {
    createHistory: createBrowserHistory_1.default,
    createHashHistory: createHashHistory_1.default,
    createMemoryHistory: createMemoryHistory_1.default
};
var useBasename_1 = require("./useBasename");
Object.defineProperty(exports, "useBasename", { enumerable: true, get: function () { return useBasename_1.default; } });
var useBeforeUnload_1 = require("./useBeforeUnload");
Object.defineProperty(exports, "useBeforeUnload", { enumerable: true, get: function () { return useBeforeUnload_1.default; } });
var useQueries_1 = require("./useQueries");
Object.defineProperty(exports, "useQueries", { enumerable: true, get: function () { return useQueries_1.default; } });
///////////////////////////////////////////////////////////////////////////////
// CONSTANTS
///////////////////////////////////////////////////////////////////////////////
var Actions_1 = require("./Actions");
Object.defineProperty(exports, "Actions", { enumerable: true, get: function () { return Actions_1.default; } });
__exportStar(require("./Actions"), exports);
